lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -513,13 +513,89 @@ struct LinIneq {
+  ///   x >= max(max(a1*x+b1 + a2*x+b2, a1*x+b1 + c2),
+  ///            max(c1 + a2*x+b2, c1 + c2))
+  ///   x >= max((a1+a2)*x+(b1+b2), a1*x+(b1+c2), a2*x+(b2+c1), c1+c2)
+  ///
+  /// Since we are combining two two-term maxima, there are four possible ways
+  /// how the terms can combine, leading to the above four-term maximum. An easy
+  /// upper bound of the form we want would be the following:
+  ///
+  ///   x >= max(max(a1+a2, a1, a2)*x + max(b1+b2, b1+c2, b2+c1), c1+c2)
+  ///
+  /// However, this is a very pessimistic upper-bound that will declare very
+  /// common patterns in the IR as unbreakable cycles, despite them being very
+  /// much breakable. For example:
+  ///
+  ///   x >= max(x, 42) + max(0, 1)  <-- breakable recursion
+  ///   x >= max(max(1+0, 1, 0)*x + max(42+0, 43, 0), 42+1)
+  ///   x >= max(x + 43, 43)         <-- unbreakable recursion!
+  ///
+  /// A better approach is to take the expanded four-term maximum, retain the
+  /// non-recursive term (c1+c2), and estimate which one of the recursive terms
+  /// (first three) will become dominant as we choose greater values for x.
+  /// Since x never is inferred to be negative, the recursive term in the
+  /// maximum with the highest scaling factor for x will end up dominating as
+  /// x tends to ∞:
+  ///
+  ///   x >= max({
+  ///              (a1+a2)*x+(b1+b2)  if (a1+a2) >= max(a1+a2, a1, a2),
+  ///              a1*x+(b1+c2)       if      a1 >= max(a1+a2, a1, a2),
+  ///              a2*x+(b2+c1)       if      a2 >= max(a1+a2, a1, a2)
+  ///            }, c1+c2)
+  ///
+  /// In case multiple cases apply, the lowest bias of the recursive term is
+  /// picked. With this, the above problematic example triggers the first or
+  /// second case and becomes:
+  ///
+  ///   x >= max((1+0)*x+(0+0), 42+1) = max(x, 43)
+  ///   x >= max(1*x+(0+1), 42+1)     = max(x+1, 43)
+  ///
+  /// Of which the first case is chosen, as it has the lower bias value.
   static LinIneq add(const LinIneq &lhs, const LinIneq &rhs) {
-    return LinIneq(lhs.rec_scale + rhs.rec_scale,
-                   lhs.rec_bias + rhs.rec_bias + lhs.nonrec_bias +
-                       rhs.nonrec_bias,
-                   0, lhs.failed || rhs.failed);
+    // Determine the maximum scaling factor among the three possible recursive
+    // terms.
+    auto scale1 = lhs.rec_scale + rhs.rec_scale; // (a1+a2)
+    auto scale2 = lhs.rec_scale;                 // a1
+    auto scale3 = rhs.rec_scale;                 // a2
+    auto bias1 = lhs.rec_bias + rhs.rec_bias;    // (b1+b2)
+    auto bias2 = lhs.rec_bias + rhs.nonrec_bias; // (b1+c2)
+    auto bias3 = rhs.rec_bias + lhs.nonrec_bias; // (b2+c1)
+    auto maxScale = std::max(scale1, std::max(scale2, scale3));
+
+    // Among those terms that have a maximum scaling factor, determine the
+    // smallest bias value.
+    Optional<int32_t> minBias = llvm::None;
+    if (scale1 == maxScale && (!minBias || bias1 < *minBias))
```

Comment:
minBias is always None here?

Reply:
Yes. I got tempted to make all thre ifs identical to emphasize the pattern and let the compiler do the work. But now on second thought that just onfuscates the code unnecessarily. Will simplify.